Refactor/remove memoization of index name prefix

JanBussieck opened the pull request:
There is an open pull request #314 concerning the issue that for multitenant applications the index prefix needs to be set dynamically (#225) which is currently not possible since the index name is memoized. I think there has been some misunderstanding concerning the point of the memoization which is not performance related, but rather a result of the fact that `index_name` is both a getter and setter (unfortunately) and it should be memoized if set explicitly.

This behavior is preserved in the changes I propose by saving the `index_name_stem` when it is set while allowing the prefix to be reevaluated every time `index_name` is accessed.

pyromaniac commented:
This code looks really cryptic, the control flow is spread all over the source. I would prefer to have less side-effects in methods.

JanBussieck commented:
You are right of course, I moved the memoization into the private method `index_name_stem`. We'll have the awkward behavior of either setting or computing a value in a single method either way, I moved this down to `index_name_stem` which is the actual value we want to memoize.

pyromaniac commented:
Alright, it looks much better. Could I ask you to add specs to consolidate this behavior? Also I'm going to think about this a little bit more over the weekend.

JanBussieck commented:
Sure will do. If you have any other suggestions concerning this change, I'd be happy to incorporate them.

JanBussieck commented:
Any idea why this little change causes so much breakage in other places?

pyromaniac commented:
Because everything else is counting on the name cache, I'm still going to think how to make it the best way, unfortunately I didn't have a chance yet :(

pyromaniac commented:
In master already
